Add an option to refresh the lock automatically

zoni opened the pull request:
This introduces a new parameter `refresh_interval` which, when set to a
value > 0 triggers the creation of a subthread that will re-set the
expire time of the lock.

This allows one to use a relatively low expires time but keep the lock
active in Redis for as long as the lock is acquired in Python code,
avoiding the stale lock issue when a Python process is killed abruptly
without a chance to release the lock in Redis.

ionelmc commented:
This is a very interesting change, I like the idea.

I have some questions/comments about functionality tho:

* Shouldn't the thread be a daemon thread (eg: so interpreter exit is not blocked by it)?
* The `thread.join` in `_stop_lock_refresher` may block up to the interval the thread uses. I think it's undesirable to have `release`/`__exit__` to wait such high amount of time.
* The `_stop_lock_refresher` method doesn't check if the thread is alive. This can easily lead to a deadlock if thread is dead (eg, you use `fork` - the new process will not have any threads).

The mongodb client also use threads for some management stuff, would be useful to check how they handle thread cleanup/termination. See https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/periodic_executor.py

jweyrich commented:
Interesting indeed. If you allow my considerations:

Would  `renew` or `renewal` make more sense than `refresh`?

> Shouldn't the thread be a daemon thread (eg: so interpreter exit is not blocked by it)?

You normally want to wait all threads before exiting. But that could be configurable, so you may get the behaviour you think is more appropriate for a given application. I'd personally prefer to join the thread, but a timeout for `join()` would be useful - if it fails, you still have the option to kill the thread, or decide to do something else.

> The thread.join in _stop_lock_refresher may block up to the interval the thread uses. I think it's undesirable to have release/__exit__ to wait such high amount of time.

Makes sense. Would use `Event.wait(interval)` so the other thread can signal it, otherwise it will just timeout.

> The _stop_lock_refresher method doesn't check if the thread is alive. This can easily lead to a deadlock if thread is dead (eg, you use fork - the new process will not have any threads).

Fork would duplicate nearly everything, though not immediately (thanks to copy-on-write). Is that bad?

ionelmc commented:
> Fork would duplicate nearly everything, though not immediately (thanks to copy-on-write). Is that bad?

Fork clones just 1 process. The threads are not cloned (but they will have broken references to them around, like the `Thread` objects).

This happens because threads in Linux are just special processes, and fork doesn't clone multiple processes.

jweyrich commented:
Right. It doesn't clone sub-processes, but still, does that sound right? I know many softwares rely on `fork` but I always felt it was the wrong approach for most cases - though it's way simpler.

ionelmc commented:
You'd think mixing threads and fork is bad, try mixing anything with signals :grin:

zoni commented:
> Shouldn't the thread be a daemon thread (eg: so interpreter exit is not blocked by it)?

It [already is](https://github.com/zoni/python-redis-lock/blob/refresh_interval/src/redis_lock/__init__.py#L131) :smiley: 

> The thread.join in _stop_lock_refresher may block up to the interval the thread uses. I think it's undesirable to have release/__exit__ to wait such high amount of time.

Agreed, that would be undesirable. I'll see about changing this behavior.

> The _stop_lock_refresher method doesn't check if the thread is alive. This can easily lead to a deadlock if thread is dead (eg, you use fork - the new process will not have any threads).

Good catch, gonna fix that too!

> The mongodb client also use threads for some management stuff, would be useful to check how they handle thread cleanup/termination

I'll take a peek, thanks for the reference.

> Would renew or renewal make more sense than refresh?

Thinking about it, yes, I think it would. I'm tempted to go for `renewal`. What do you think @ionelmc?

ionelmc commented:
> Thinking about it, yes, I think it would. I'm tempted to go for renewal. What do you think @ionelmc?

How about "keepalive"?  Or "keepalive_interval"? Needs to sound a bit dangerous. Threads are involved. :grin:

zoni commented:
*keepalive* still doesn't sound that dangerous to me :wink:. I personally like *renewal* more but *keepalive* covers the functionality as well so I'm happy with either option.

zoni commented:
In the end I decided to go with *"renewal"* over *"keepalive"*, mainly because renewal seems more commonly used in other software in this context as well (if you strongly prefer *keepalive* though, I'll happily change it again).

Also note that the docstring for this parameter already mentions a thread is used for the lock renewing.

On a more technical level, I believe I've now addressed the concerns that were raised.

zoni commented:
The build failure is unrelated to the PR, it is failing in `tests/test_redis_lock.py::test_no_overlap`, which looks like it fails very sporadically (seen it fail once before as well).

ionelmc commented:
One more question: does it make sense to have a "renewal interval"? 

I think it just opens the door to user-errors and confusion:

* The user can set a renewal duration that's equal or greater than the expiry - thus causing the whole thing to malfunction.
* The default (`-1`) may be confusing - is it different than `None`, `False` or `0`?

What if `renewal` was just a boolean flag? And we'd pick a sensible value for the interval?

What would be a sensible value? 

* `expiry / 2`? This may be wasteful, plus it causes overlap.
* `expiry - 1 second`? This implies awareness of latency to the redis server. Maybe this assumes too much. Plus it can't work work for sub-second situations.
* `expiry * 2 / 3`? This solves the overlap ... maybe this is better?

Then there would be 2 options (if we want allow users to customize interval):

* `renewal` - `True` or `False`
* `renewal_interval` - `None` (default: it will mean the computed `2/3` or whatever we think is best) or an integer.

Maybe the `renewal_interval` could even be  a *hidden* option - eg, you have to subclass the `Lock` class to customize it?

What do you think? I think we should make the obvious choices for the user in the tipical scenario, instead of letting the user wondering what's best and make bad choices.

vdloo commented:
Just tested this with our setup and it doesn't work yet. If the program with the lock crashes and you run it again before the expire, it will block forever [here](https://github.com/zoni/python-redis-lock/blob/refresh_interval/src/redis_lock/__init__.py#L93) because the expire is 0. 

You can fix that by changing redis_lock/\__init__.py:[93](https://github.com/zoni/python-redis-lock/blob/refresh_interval/src/redis_lock/__init__.py#L93) to
```
self._client.blpop(self._signal, self._expire or self._lock_renewal_interval or 0)
```

and redis_lock/\__init__.py:[90](https://github.com/zoni/python-redis-lock/blob/refresh_interval/src/redis_lock/__init__.py#L90)
```
busy = not self._client.set(self._name, self._id, nx=True, ex=self._expire or self._lock_renewal_interval)
```

also I agree on the renewal=-1, it doesn't feel like the pythonesque thing to do. None would be better.

zoni commented:
> Just tested this with our setup and it doesn't work yet. If the program with the lock crashes and you run it again before the renewal runs out, it will block forever here because the expire is 0. 

I don't see how this could be the case, unless you explicitly set expire to 0/None, in which case having a renewal doesn't make sense. 

The call to [self._client.set()](https://github.com/zoni/python-redis-lock/blob/refresh_interval/src/redis_lock/__init__.py#L90) definitely should ***not*** have `ex=self._expire or self._lock_renewal_interval` but should remain at `ex=self._expire` because it sets the expire time for the lock, in no case should it set the renew interval as the expiry time.

We could however consider enforcing that an `expire` time is supplied when `renewal` is given, as it generally doesn't make sense to set the latter without the former.

zoni commented:
> One more question: does it make sense to have a "renewal interval"?

I'd considered the same thing (and picked `expiry * 2 / 3` as the best sensible default as well) but eventually decided not to to keep the code more simple/predictable. If you're selecting an expiry time, it's a small thing to think of a suitable renewal interval as well. But I like the idea of just making `renewal` a boolean and having a separate `renewal_interval` you can set to override the default computed value.

zoni commented:
I've changed the parameter to `auto_renewal` now, a boolean flag that enables or disables the behavior (it's off by default). I figured this expresses the *"magic"* threaded stuff some more :smiling_imp:.

It picks a renewal interval itself, `expiry*2/3` as was already suggested. If you want a different interval you'll have to subclass the `Lock`, but it's explained in the documentation for the parameter so it's clear to users who have that need.

As part of this, I made it required to set an `expire` when setting `auto_renewal` because without an expire we have no way of calculating a suitable interval.

zoni commented on src/redis_lock/__init__.py:
```diff
@@ -29,15 +30,49 @@ class NotAcquired(RuntimeError):
+        """
         assert isinstance(redis_client, StrictRedis)
+        if auto_renewal and expire is None:
+            raise ValueError("Expire may not be None when auto_renewal is set")
```
Maybe `TypeError` is more suitable than `ValueError` here?

TypeError =  Inappropriate argument type.
ValueError = Inappropriate argument value (of correct type).

zoni commented:
@ionelmc, do you have any more thoughts and/or comments on the current implementation?

ionelmc commented on README.rst:
```diff
@@ -139,11 +139,19 @@ You would write your functions like this::
+`auto_renewal` parameter in combination with `expire` to set a time-out on the lock, but let `Lock()` automatically
+keep resetting the expire time while your application code is executing::
+
+    # Expire the lock after 60 seconds, but keep renewing it every 30 seconds
```
This should state 40 seconds.

ionelmc commented:
@zoni Looks good. Can you fix that typo? I'll merge after that.

zoni commented:
Done! I removed a mention of that interval entirely as we made it a behind-the-scenes thing that users shouldn't normally need to concern themselves with.

ionelmc commented:
Woohoo! Thanks!

vdloo commented:
great! can you guys make a release to https://pypi.python.org/pypi/python-redis-lock ?

ionelmc commented:
Released 2.2.0
